### PR TITLE
ISPN-13463 Expose JGroups metrics

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-13463

The first commit reorganizes the build a bit to move all build-related modules to a common subdirectory

### CODE REVIEW

**Comment by tristantarrant** [comment]
Updated with support for xsite metrics

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java [hunk 1]
```diff
@@ -468,6 +476,7 @@ public void start() {
       setXSiteViewListener(channelCallbacks);
 
       startJGroupsChannelIfNeeded();
+      metricsRegistration.registerMetrics(channel, metricsCollector.wired());
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:479]: I think you should keep track of `MetricID` and unregister them on stop/disconnect.

Comment (tristantarrant) [reply core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:479]: Done

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java [hunk 2]
```diff
@@ -159,23 +166,35 @@
...
-   @Inject @ComponentName(KnownComponentNames.NON_BLOCKING_EXECUTOR)
+   @Inject
+   @ComponentName(KnownComponentNames.NON_BLOCKING_EXECUTOR)
    protected ExecutorService nonBlockingExecutor;
-   @Inject protected CacheManagerJmxRegistration jmxRegistration;
-   @Inject protected GlobalXSiteAdminOperations globalXSiteAdminOperations;
+   @Inject
+   protected CacheManagerJmxRegistration jmxRegistration;
+   @Inject
+   protected GlobalXSiteAdminOperations globalXSiteAdminOperations;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:189]: Would be great if we could prevent the reformatting here as it is creating quite a bit of additional noise.

File: core/src/main/java/org/infinispan/metrics/impl/MetricsCollector.java [hunk 3]
```diff
@@ -181,8 +181,8 @@ private static String generateRandomName() {
       return metricIds;
    }
 
-   public void unregisterMetric(MetricID metricId) {
-      boolean removed = registry.remove(metricId);
+   public void unregisterMetric(Object metricId) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/metrics/impl/MetricsCollector.java:184]: Is this change really necessary? I saw that it seems to be using MetricID everywhere and if not this change is setting ourselves up for a ClassCastException with no compiler checks :(

Comment (tristantarrant) [reply core/src/main/java/org/infinispan/metrics/impl/MetricsCollector.java:184]: MetricID is a smallrye-metrics implementation detail which should not leak everywhere. Since we will be migrating to micrometer, we can handle this as an opaque type

Comment (wburns) [reply core/src/main/java/org/infinispan/metrics/impl/MetricsCollector.java:184]: Okay, I just worry because the line below does a cast, so we lose that type info but still rely upon the concrete class. I guess you are doing this for an upcoming change though.

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java (hunk 2 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java]: Cleaned it up.

File: core/src/main/java/org/infinispan/metrics/impl/MetricsCollector.java (hunk 3 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/metrics/impl/MetricsCollector.java:184]: Yeah, the task is https://issues.redhat.com/browse/ISPN-13420

**Comment by pruivo** [comment]
@tristantarrant the root `pom.xml` and checkstyle `pom.xml` are broken (the parent needs to be updated). If you delete `~/.m2/repository` you will notice :)

**Review by pruivo — APPROVED** [review]

**Comment by tristantarrant** [comment]
CI looks ok

**Comment by pruivo** [comment]
merged! thanks @tristantarrant !

**Comment by pruivo** [comment]
are you going to create a PR for 13.0.x? :fearful:
